src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -59,12 +62,19 @@ ss::future<> refresh_credentials::do_start() {
     return ss::do_until(
       [this] { return _gate.is_closed() || _as.abort_requested(); },
       [this] {
-          return fetch_and_update_credentials().handle_exception_type(
-            [](const ss::sleep_aborted& ex) {
+          return fetch_and_update_credentials()
+            .handle_exception_type([](const ss::sleep_aborted& ex) {
                 vlog(
                   clrl_log.info,
                   "stopping refresh_credentials loop: {}",
                   ex.what());
+            })
+            .handle_exception_type([this](const ss::timed_out_error& err) {
```

Comment:
Shouldn't we also handle any unknown exceptions here?  e.g. hypothetically if there was some other exception we hadn't thought about, we would rather keep on with the do_until than silently stop

Reply:
There is actually code to handle retrying of errors in https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_roles/request_response_helpers.cc#L48 which should catch and convert exceptions to domain errors which are handled at the top level in `refresh_credentials.cc`. 

The timeout error was not caught there though, and while looking at the code for exception handling I see that there is a bug in this `do_request` method in the helper `request_response_helpers.cc`. 

It returns futures from an expression (no coroutines) and tries to catch exceptions around the call, which does not work as expected. 

It is also the reason we did not have warning logs in grafana from the catch blocks. We either need to change the function to handle errors in a continuation attached to the future in `do_request` or use coroutines. I will change it to use the latter. 

With this change it should no longer be required to do `handle_exception_type` at the top level. 

Thanks for pointing this out.

Reply:
fixed to handle exceptions with coroutines